### DISCUSSION TITLE
How to update odoo source code after running the instance

### DISCUSSION THREAD

**Discussion by AungKoKoLin1997** [Tecnativa/doodba#635] [Q&A]
I am using doodba and I am testing to use for production. So, I set up the server and my instance is running well.
Then, I want to update my odoo source code.
So, I run the invoke `git-aggregate` and this is the log.

```
lin@localhost:~/odoo18/odoo/custom/src$ invoke git-aggregate
Reinitialized existing Git repository in /home/lin/odoo18/.git/
pre-commit installed at .git/hooks/pre-commit
time="2024-11-06T04:29:25Z" level=warning msg="/home/lin/odoo18/setup-devel.yaml: the attribute `version` is obsolete, it will be ignored, please remove it to avoid potential confusion"
 Network odoo18_public  Creating
 Network odoo18_public  Created
time="2024-11-06T04:29:25Z" level=warning msg="Found orphan containers ([odoo18_db_1]) for this project. If you removed or renamed this service in your compose file, you can run this command with the --remove-orphans flag to clean it up."
doodba INFO: Running gitaggregate with /opt/odoo/custom/src/repos.yaml
(INFO) [04:29:26] git_aggregator.repo  odoo   Start aggregation of /opt/odoo/custom/src/odoo
(INFO) [04:29:26] git_aggregator.repo  odoo   Switch to branch 18.0
Reset branch '18.0'
Your branch is up to date with 'origin/18.0'.
(INFO) [04:29:27] git_aggregator.repo  odoo   Remote already exists ocb <https://github.com/OCA/OCB.git>
(INFO) [04:29:27] git_aggregator.repo  odoo   Remote already exists odoo <https://github.com/odoo/odoo.git>
(INFO) [04:29:27] git_aggregator.repo  odoo   Remote already exists openupgrade <https://github.com/OCA/OpenUpgrade.git>
(INFO) [04:29:27] git_aggregator.repo  odoo   Fetching required remotes
From https://github.com/OCA/OCB
 * branch              18.0       -> FETCH_HEAD
(INFO) [04:29:28] git_aggregator.repo  odoo   Checking repo status
(INFO) [04:29:28] git_aggregator.repo  odoo   Reset branch to ocb 18.0
(INFO) [04:29:29] git_aggregator.repo  odoo   Execute shell after commands
(INFO) [04:29:29] git_aggregator.repo  odoo   End aggregation of /opt/odoo/custom/src/odoo
```

I thought it was looking good and applied the latest source code to the container. But when I check git log of odoo directory inside container, it is not the latest commit and seem like not working. Should `invoke git-aggregate` only use for development environment?
Which way do I have to use to update my source code?

**Comment by ap-wtioit** [comment]
When you want to update the image for your prod instance you need to build the image with
* `invoke img-build` (make sure prod.yaml is symlinked as docker-compose.yaml)

or
* `docker compose -f prod.yaml build odoo`

After this the odoo image is now a prod image. To get a develop image again your have to build the dev image again. So this is best done on a CI environment.

**Reply by AungKoKoLin1997** [reply]
I build the image with

- `docker compose -f prod.yaml build odoo`.

After that, I run `invoke git-aggregate`. Then, restart with
`sudo docker compose restart odoo`

Then, I go to my odoo container and when I checked the git log of odoo addons directory(`/opt/odoo/custom/src/odoo`), it is not still latest.

<img width="1435" alt="Screenshot 2024-11-06 at 14 16 14" src="https://github.com/user-attachments/assets/995988eb-e893-4b04-9c94-4c3bf0b93bfa">

Am I missing some steps?

**Reply by pedrobaeza** [reply]
`invoke git-aggregate` is only for development environment, not for production. The source code you have downloaded doesn't server at all.

On production environment, once you have build the image, you have to "use it". For replacing the current running image with the new one, do `docker compose up -d`. `restart` command is still using the same old image. This is something about Docker though, not Doodba itself. For auto-updating changed modules, run `docker compose run --rm -l traefik.enable=false odoo click-odoo-update`.

**Reply by ap-wtioit** [reply]
* `docker compose -f prod.yaml build odoo` is for prod environment
* `invoke git-aggregate` is for dev environment (it switches docker-compose.yaml back to devel.yaml)

i would not expect to use both in the same environment

for prod environment
* `docker compose -f prod.yaml build odoo` (or `invoke img-build` with symlink to prod.yaml)
* `docker compose -f prod.yaml up -d odoo` (or `invoke start` with symlink to prod.yaml)

for dev environment:
https://github.com/Tecnativa/doodba-copier-template/blob/main/docs/daily-usage.md#development

depending on your repos.yaml and addons.yaml the outcome might be a strange merge (even if i don't see that in your provided logs)

also one time we had somewone accidentially messing with git and doing something like:
```
git checkout -b ocb/18.0
```
which could also explain why ocb 18.0 no longer updates on your side.

 if you have pushed all your odoo changes your could try to delete odoo/custom/src/odoo and start over again (**DANGER** do not do this if you have important local changes)
